File: cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int K) {
        int[][] g = new int[n][n];
        for (int[] f : flights) {
            g[f[0]][f[1]] = f[2];
        }

        // (1) adding the distance array to track min distance to each node
        int[] distance = new int[n];
        Arrays.fill(distance, Integer.MAX_VALUE / 2);

        //  PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        // (2) change heap to normal queue
        Queue<int[]> heap = new LinkedList<>();

        heap.offer(new int[]{0, src, K + 1});
        distance[src] = 0;

        while (!heap.isEmpty()) {
            int[] cur = heap.poll();
            int price = cur[0], place = cur[1], remainStops = cur[2];
            
          //  if (place == dst) return price;
            
            if (remainStops > 0) {
                for (int i = 0; i < n; i++) {
                    if (g[place][i] > 0 && distance[i] >= price + g[place][i]) {                        
                        heap.offer(new int[]{price + g[place][i], i, remainStops - 1});
                        distance[i] = price + g[place][i];
                    }
                }
            }
        }
        
        return distance[dst] == Integer.MAX_VALUE / 2 ? -1 : distance[dst];
    }
}

File: cheapest-flights-within-k-stops/cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        if src==dst:
            return 0
        graph = collections.defaultdict(list)
        for cur, to, price in flights:
            graph[cur].append((to,price))
        queue = deque([(-1,src,0)])  # (到当前结点的步数step，当前结点cur，到当前结点的累计价钱)
        
        
        # price = {src:0}  #从起点到当前结点的距离
        # price = collections.defaultdict(lambda: float('inf')) # 距离哈希表
        price = [float('inf') for _ in range(n)]
        price[src]=0
        
        
        while queue:
            step, cur, cum_price=queue.popleft()
            if step>=k:   #如果k==K， 已经走了k步了，不能再从这个结点出发走下一步了
                continue
            if cur == dst: #pos==dst:并不是到达终点就可以收工，可能还有更小的price 存在
                continue
            for nextnode, price_tonext in graph[cur]:#遍历当前结点能去的下一节点 
                temp = price_tonext+cum_price #如果从当前结点出发 到 下一节点 的总路程 temp
                if  price[nextnode] > temp :  #如果temp 比原有到下一节点的价钱更少， 
                    price[nextnode]= temp     #那就更新到下一节点的价钱
                    queue.append((step+1, nextnode, price[nextnode]))

        return price[dst] if price[dst]!=float('inf') else -1